### PR TITLE
[HDF5] Update - 2 - Container component IO

### PR DESCRIPTION
**Pull request by sunethwarna**
**📝 Description**
This PR replaces all the `#pragma omp` calls with the `ParallelUtilities`, and unifies the behaviour between all the component IOs.

Removes the dependency on `RegisteredComponentLookup` by replacing its uses with more clear fold expressions.

** TODO **

- [x] Add zero data rank tests for component IO.

**🆕 Changelog**
- Update `ComponentDataIO`.
- Updated tests.

### CODE REVIEW

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+{
+public:
+    ///@name Type definitions
+    ///@{
+
+    template<class TDataType>
+    using TLSType = char;
+
+    template<class TDataType>
+    using ComponentDataType = char;
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]: Too much `Data` in type names

```suggestion
    using ComponentType = char;
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+
+    ///@}
+    ///@name Public operations
+    ///@{
+
+    template<class TEntityType>
+    inline char GetValue(
+        const TEntityType& rEntity,
+        const Flags& rFlag,
+        char& rTLS) const
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h:56]: I'm not sure what `TLSType` is for here, but I guess this is one of them

```suggestion
        TLSType<TEntityType>& rTLS) const
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+    ///@}
+
+private:
+    ///@name Private member variables
+    ///@{
+
+    const double mAlphaBossak;
+
+    ///@}
+};
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]: Why does Bossak need a whole separate infrastructure?

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+        const Variable<TDataType>& rVariable,
+        const TDataType& rValue) const
+    {
+        KRATOS_ERROR << "VertexValueIO does not support setting values from vertices";
+    }
+
+    ///@}
+};
+
+class GaussPointValueIO
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]:
```suggestion
class GaussPointIO
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+    ///@{
+
+    template<class TEntityType, class TDataType>
+    inline const ComponentDataType<TDataType>& GetValue(
+        const TEntityType& rEntity,
+        const Variable<TDataType>& rVariable,
+        TLSType<TDataType>& rTLS) const
+    {
+        // Since the array_1d<double, 4> and array_1d<double 9> interfaces are not present in
+        // elements and conditions.
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]: should we define them then?

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+    ///@name Private member variables
+    ///@{
+
+    const ProcessInfo& mrProcessInfo;
+
+    ///@}
+};
+
+template<class TContainerIOType>
+std::string GetContainerIOType()
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]:
```suggestion
std::string GetContainerIOName()
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+        return "INTERPOLATED";
+    } else if constexpr(std::is_same_v<TContainerIOType, GaussPointValueIO>) {
+        return "GAUSS_POINT";
+    } else {
+        static_assert(!std::is_same_v<TContainerIOType, TContainerIOType>, "Unsupported container io type.");
+    }
+}
+
+template<class TContainerType>
+std::string GetContainerType()
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]:
```suggestion
std::string GetContainerName()
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 8]
```diff
@@ -0,0 +1,477 @@
...
+        return rModelPart.GetCommunicator().LocalMesh().Conditions();
+    } else if constexpr(std::is_same_v<TContainerType, ModelPart::ElementsContainerType>) {
+        return rModelPart.GetCommunicator().LocalMesh().Elements();
+    } else {
+        static_assert(!std::is_same_v<TContainerType, TContainerType>, "Unsupported container type.");
+    }
+}
+
+template<class TContainerType, class TComponentType, class TContainerDataIO, class TDataType>
+void CopyToContiguousDataArray(
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]:
```suggestion
void CopyToContiguousArray(
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 9]
```diff
@@ -0,0 +1,477 @@
...
+    const TContainerType& rContainer,
+    const TComponentType& rComponent,
+    const TContainerDataIO& rContainerDataIO,
+    TDataType* pBegin)
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]: can you please also pass the array's size as well? It'll make life much easier if we could validate that the array is big enough to fit all entries, instead of segfaulting.

```suggestion
    const TContainerType& rContainer,
    const TComponentType& rComponent,
    const TContainerDataIO& rContainerDataIO,
    TDataType* pBegin,
    const std::size_t Size)
```

If this pattern occurs frequently, we could also add a `Span` (similar to [`std::span`](https://en.cppreference.com/w/cpp/container/span)) class that models a view over a contiguous array. It's a handy, safe and lightweight thing to pass around compared to just juggling with raw pointers.

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 10]
```diff
@@ -0,0 +1,477 @@
...
+{
+    KRATOS_TRY
+
+    using component_data_type = typename Internals::template ComponentTraits<TComponentType>::ValueType;
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]:
```suggestion
    using component_type = typename Internals::template ComponentTraits<TComponentType>::ValueType;
```

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 11]
```diff
@@ -0,0 +1,477 @@
...
+
+    using value_type = typename TContainerDataIO::template ComponentDataType<component_data_type>;
+
+    using value_type_traits = DataTypeTraits<value_type>;
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h:406]: `value` or `data`? The naming should be consistent. I'm voting for `value` in this case because it distinguishes between the `value_type` stored in an array and `component_type` which might be identical to `value_type` or compose it, while `data` doesn't say much.

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 12]
```diff
@@ -0,0 +1,477 @@
...
+    const auto& initial_value = rContainerDataIO.GetValue(rContainer.front(), rComponent, tls);
+
+    // get the stride from the first element to support dynamic types.
+    const auto stride = value_type_traits::Size(initial_value);
+
+    IndexPartition<unsigned int>(rContainer.size()).for_each(tls, [&rContainer, &rComponent, &rContainerDataIO, pBegin, stride](const auto Index, auto& rTLS) {
+        const auto& value = rContainerDataIO.GetValue(*(rContainer.begin() + Index), rComponent, rTLS);
+        TDataType const* p_value_begin = value_type_traits::GetContiguousData(value);
+        auto p_array_start = pBegin + Index * stride;
+        std::copy(p_value_begin, p_value_begin + stride, p_array_start);
```
Comment (matekelemen) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]:
```suggestion
        auto p_subrange_begin = pBegin + Index * stride;
        std::copy(p_value_begin, p_value_begin + stride, p_subrange_begin);
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 13]
```diff
@@ -40,13 +40,15 @@ namespace HDF5
 ///@{
 
 /// A class for IO of element data in HDF5.
-template <typename TContainerType, typename TContainerItemType, typename... TComponents>
+template <class TContainerType, class TContainerDataIO, class... TComponents>
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h]:
```suggestion
template <class TContainerType, class TContainerIO, class... TComponents>
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 14]
```diff
@@ -55,43 +57,72 @@ class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
...
     ///@name Private Operations
     ///@{
 
-    template <typename... Targs>
-    void WriteRegisteredComponent(const std::string& rComponentName, Targs&... args);
-
-    template <typename... Targs>
-    void ReadRegisteredComponent(const std::string& rComponentName, Targs&... args);    
+    template<class TComponentType>
+    bool WriteComponentData(
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h]:
```suggestion
    bool WriteComponents(
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 15]
```diff
@@ -55,43 +57,72 @@ class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
...
+        const std::string& rComponentName,
+        const TContainerDataIO& rContainerDataIO,
+        const TContainerType& rLocalContainer,
+        Parameters Attributes,
+        WriteInfo& rInfo);
+
+    template<class TComponentType>
+    bool ReadComponentData(
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h]:
```suggestion
    bool ReadComponents(
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 16]
```diff
@@ -40,13 +40,15 @@ namespace HDF5
 ///@{
 
 /// A class for IO of element data in HDF5.
-template <typename TContainerType, typename TContainerItemType, typename... TComponents>
+template <class TContainerType, class TContainerDataIO, class... TComponents>
 class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h:44]: This is probably the most important class in the entire application, and yet it's not documented. We need detailed docs here about what this class does.

```suggestion
class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 17]
```diff
@@ -55,43 +57,72 @@ class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
     ///@{
 
     /// Constructor.
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h:59]: ... and I'm not talking about these kinds of docs. Please explain each argument, especially what's expected in `Settings`.

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 18]
```diff
@@ -55,43 +57,72 @@ class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
...
+        const ModelPart& rModelPart,
+        const TContainerDataIO& rContainerDataIO,
+        const Parameters Attributes);
 
-protected:
-    ///@name Protected Member Variables
-    ///@{
+    std::map<std::string, Parameters> Read(
+        ModelPart& rModelPart,
+        const TContainerDataIO& rContainerDataIO);
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h]: missing docs about what this actually returns

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h [hunk 19]
```diff
@@ -55,43 +57,72 @@ class KRATOS_API(HDF5_APPLICATION) ContainerComponentIO
...
 
     ///@}
     ///@name Operations
     ///@{
 
-    ///@}
+    void Write(
+        const ModelPart& rModelPart,
+        const TContainerDataIO& rContainerDataIO,
+        const Parameters Attributes);
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.h]: missing docs about what's expected in `Attributes`

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 20]
```diff
@@ -1,384 +1,91 @@
...
-template <typename TComponent>
-class ReadElementComponent
-    : public ContainerItemComponentIO<ElementType>::ReadComponentFunctor<TComponent, ComponentElementSynchronizingFunctor>
-{
-};
-
-template <typename TComponent>
-class WriteConditionComponent
-    : public ContainerItemComponentIO<ConditionType>::WriteComponentFunctor<TComponent, void>
+namespace NewContainerComponentIOUtilities
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp:35]:
```suggestion
// TODO: rename this namespace once HDF5 refactoring is complete
namespace NewContainerComponentIOUtilities
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 21]
```diff
@@ -1,384 +1,91 @@
...
     : mpFile(pFile)
 {
     KRATOS_TRY;
 
     Parameters default_params(R"(
         {
-            "prefix": "",
+            "prefix"           : "",
             "list_of_variables": []
         })");
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp:106]: this should be in a `GetDefaultParameters` static member function

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 22]
```diff
@@ -1,384 +1,91 @@
...
 
     Settings.AddMissingParameters(default_params);
 
-    mComponentPath = Settings["prefix"].GetString() + rComponentPath;
+    mComponentPrefix = Settings["prefix"].GetString() + rLegacySuffix;
     mComponentNames = Settings["list_of_variables"].GetStringArray();
 
+    KRATOS_ERROR_IF(mComponentPrefix == "" || mComponentPrefix == "/")
+        << "The prefix should not be blank or \"/\" [ prefix = " << mComponentPrefix << " ].\n";
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]:
```suggestion
        << "The prefix must not be empty or \"/\" [ prefix = " << mComponentPrefix << " ].\n";
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 23]
```diff
@@ -389,477 +96,331 @@ ContainerComponentIO<TContainerType, TContainerItemType, TComponents...>::Contai
...
+    KRATOS_TRY
+
+    if constexpr(
+        std::is_same_v<TContainerType, NodesContainerType> ||
+        std::is_same_v<TContainerType, ConditionsContainerType> ||
+        std::is_same_v<TContainerType, ElementsContainerType>) {
+        return Read(Internals::GetLocalContainer<TContainerType>(rModelPart), rContainerDataIO, rModelPart.GetCommunicator());
+    } else {
+        KRATOS_ERROR << "Unsupported container type.";
+        return std::map<std::string, Parameters>{};
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]:
```suggestion
        return {};
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 24]
```diff
@@ -389,477 +96,331 @@ ContainerComponentIO<TContainerType, TContainerItemType, TComponents...>::Contai
...
-                                                                             Targs&... args)
-{
-    RegisteredComponentLookup<Variable<array_1d<double, 3>>, Variable<double>, Variable<int>,
-                              Variable<Vector<double>>, Variable<Matrix<double>>>(rComponentName)
-        .Execute<ReadConditionComponent>(args...);
-}
+    // Write each variable.
+    for (const auto& r_component_name : mComponentNames) {
+        const bool is_component_written = (... || (WriteComponentData<TComponents>(
+                        r_component_name, rContainerDataIO, rLocalContainer, Attributes.Clone(), info)));
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: not sure about this one, but shouldn't it be `&&`?


```suggestion
        const bool is_component_written = (... && (WriteComponentData<TComponents>(
                        r_component_name, rContainerDataIO, rLocalContainer, Attributes.Clone(), info)));
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 25]
```diff
@@ -389,477 +96,331 @@ ContainerComponentIO<TContainerType, TContainerItemType, TComponents...>::Contai
...
-template <>
-template <class ...TArguments>
-void ContainerComponentIO<VertexContainerType, Vertex, Flags>::WriteRegisteredComponent(
-    const std::string& rComponentName, TArguments&... rArguments)
-{
-    RegisteredComponentLookup<Flags>(rComponentName).Execute<WriteVertexComponent>(rArguments...);
-}
+    if (mComponentNames.size() == 0) {
+        return attributes;
+    }
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: my MPI spider senses are tingling. Can a situation ever occur when `mComponentNames` is not consistent across ranks?

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 26]
```diff
@@ -389,477 +96,331 @@ ContainerComponentIO<TContainerType, TContainerItemType, TComponents...>::Contai
...
         }
-    }
 
-    std::vector<TContainerItemType*> local_items;
-    GetContainerItemReferences(local_items, rContainer);
-    std::size_t start_index, block_size;
-    std::tie(start_index, block_size) = StartIndexAndBlockSize(*mpFile, mComponentPath);
+        // add the shape to attributes.
+        KRATOS_ERROR_IF(Attributes.Has("__data_shape"))
+            << "The reserved keyword \"__data_shape\" is found. Please remove it from attributes.";
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]:
```suggestion
            << "reserved keyword \"__data_shape\" is found. Please remove it from the attributes.";
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp [hunk 27]
```diff
@@ -389,477 +96,331 @@ ContainerComponentIO<TContainerType, TContainerItemType, TComponents...>::Contai
...
-                       Vector<int> const& rData,
-                       std::vector<TContainerItemType*>& rContainerItems)
-{
-    KRATOS_TRY;
+        if (attributes.Has("__data_dimension")) attributes.RemoveValue("__data_dimension");
+        if (attributes.Has("__data_shape")) attributes.RemoveValue("__data_shape");
+        if (attributes.Has("__container_type")) attributes.RemoveValue("__container_type");
+        if (attributes.Has("__data_name")) attributes.RemoveValue("__data_name");
+        if (attributes.Has("__data_location")) attributes.RemoveValue("__data_location");
+        if (attributes.Has("__mesh_location")) attributes.RemoveValue("__mesh_location");
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: these should be in some central location so if we ever need to update them, we can do so more easily (and they'd also be a bit easier to look through), Also, `ContainerComponentIO`'s docs should mention these reserved attributes.

File: applications/HDF5Application/custom_io/hdf5_file.cpp [hunk 28]
```diff
@@ -795,7 +795,9 @@ bool File::HasDataType(const std::string& rPath) const
     if constexpr(std::is_same_v<TDataType, int>) {
         return type == H5T_INTEGER;
     } else if constexpr(std::is_same_v<TDataType, double>) {
-        return type == H5T_FLOAT;
+        return (type == H5T_FLOAT);
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_file.cpp]:
```suggestion
        return type == H5T_FLOAT;
```

File: applications/HDF5Application/custom_io/hdf5_file.cpp [hunk 29]
```diff
@@ -795,7 +795,9 @@ bool File::HasDataType(const std::string& rPath) const
...
+    } else if constexpr(std::is_same_v<TDataType, char>) {
+        return (type == H5T_INTEGER);
```
Comment (matekelemen) [inline applications/HDF5Application/custom_io/hdf5_file.cpp]:
```suggestion
        return type == H5T_INTEGER;
```

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 24 above)
Comment (matekelemen) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: nevermind I think I understand now ...

this loops over the types and checks whether the name is registered in the corresponding instantiation of `KratosComponents`. Most return `false` and only one should return `true` (for the type of the requested variable). Is that correct?

File: applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp [hunk 30]
```diff
@@ -10,189 +10,122 @@
...
-
-void VertexContainerCoordinateIO::WriteWithoutIDs(const Detail::VertexContainerType& rVertices)
+void VertexContainerCoordinateIO::WriteWithoutIDs(
+    const Detail::VertexContainerType& rVertices,
+    Parameters Attributes)
 {
     KRATOS_TRY
 
-    const std::string path = mComponentPath + "/POSITION";
+    const std::string path = mPathPrefix + "/POSITION";
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp:94]: Perhaps filesystem::path would be better here

File: applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp [hunk 31]
```diff
@@ -10,189 +10,122 @@
...
-
-    KRATOS_CATCH("");
-}
-
-
-VertexContainerCoordinateIO::VertexContainerCoordinateIO(Parameters parameters,
-                                                         File::Pointer pFile)
-    : VertexContainerIO(VertexContainerCoordinateIO::FormatParameters(parameters), pFile)
+void VertexContainerCoordinateIO::Write(
+    const Detail::VertexContainerType& rVertices,
```
Comment (philbucher) [inline applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp:47]: Why vertex? Usually we use node

File: applications/HDF5Application/custom_python/add_custom_io_to_python.cpp [hunk 32]
```diff
@@ -29,45 +28,192 @@
...
+
+    using data_io = typename TContainerComponentIOWrapper::ContainerIOType;
+
+    using container_type = typename TContainerComponentIOWrapper::ContainerType;
 
+    using container_data_io_type = typename TContainerComponentIOWrapper::ContainerDataIOType;
+
+    std::string container_name, arg_name;
+    if constexpr(std::is_same_v<container_type, ModelPart::NodesContainerType>) {
+        container_name = "Nodal";
```
Comment (philbucher) [inline applications/HDF5Application/custom_python/add_custom_io_to_python.cpp:94]: Nodal is inconsistent with Element and Condition

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 33]
```diff
@@ -0,0 +1,477 @@
...
+std::string GetContainerType()
+{
+    if constexpr(std::is_same_v<TContainerType, ModelPart::NodesContainerType>) {
+        return "NODES";
+    } else if constexpr(std::is_same_v<TContainerType, ModelPart::ConditionsContainerType>) {
+        return "CONDITIONS";
+    } else if constexpr(std::is_same_v<TContainerType, ModelPart::ElementsContainerType>) {
+        return "ELEMENTS";
+    } else if constexpr(std::is_same_v<TContainerType, Detail::VertexContainerType>) {
+        return "VERTICES";
```
Comment (philbucher) [inline applications/HDF5Application/custom_utilities/container_io_utils.h:358]: Nodes would be more consistent

File: applications/HDF5Application/custom_utilities/container_io_utils.h [hunk 34]
```diff
@@ -0,0 +1,477 @@
...
+        if (rVariable == ACCELERATION) {
+            return (1.0 - mAlphaBossak) * rNode.FastGetSolutionStepValue(rVariable, 0) + mAlphaBossak * rNode.FastGetSolutionStepValue(rVariable, 1);
+        } else {
+            return rNode.FastGetSolutionStepValue(rVariable);
+        }
+    }
+
+    template<class TDataType>
+    inline void SetValue(
+        ModelPart::NodeType& rNode,
```
Comment (philbucher) [inline applications/HDF5Application/custom_utilities/container_io_utils.h]: Just Node&

File: applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp (hunk 30 above)
Comment (matekelemen) [reply applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp:94]: that wouldn't work on windows because it would translate to `...\\POSITION`, but this is a path within HDF5

File: applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp (hunk 31 above)
Comment (matekelemen) [reply applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp:47]: I'm to blame for this one. I came up with [`Vertex`](https://github.com/KratosMultiphysics/Kratos/blob/358979aaeb2b2d82bf7dd63992f48e984f6a04fb/applications/HDF5Application/custom_utilities/vertex.h#L31) in the context of point output to distinguish between `Node`s and general points within the mesh.

A `Vertex` is basically a point inside an element that interpolates nodal values.

File: applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp (hunk 30 above)
Comment (philbucher) [reply applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp:94]: I think you can also get the path with /

**Comment by matekelemen** [comment]
@sunethwarna btw how's the plan on adding MPI tests with empty model parts doing? These cases have always been the bane of the app.

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 1 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 2 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h:56]: Since it is always char, I don't see a difference in using `char` over `TLSType<TEntityType>`, apart from the fact use of templates makes my head go crazy someties ;)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 3 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: The way transient adjoints were done originally by using this modified output scheme of adjoints because, at the early days of transient adjoints, there were so little room for transient data passing. I would keep this for time being, until the old `VMSAdjointElement` is retired. [It is the only place this is required for fluid transient adjoint computations]

What this does is, it stores the adjoint specific `ACCELERATION` value as `ACCELERATION` in the HDF5 otuput, so when the old `VMSAdjointElement` reads `ACCELERATION`, it gets the adjoint sepcific `ACCELERATION` instead of the actual `ACCELERATION`. I understand this is workaround, we had in early days but for now, lets keep it.

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 4 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 5 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: I have done it in PR #11743  :).... now we don't need them :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 6 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 7 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 8 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 10 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 12 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h (hunk 14 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.h]: done :)

File: applications/HDF5Application/custom_io/hdf5_container_component_io.h (hunk 15 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.h]: done :)

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 24 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: You are correct 👍

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 22 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: done :)

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 21 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp:106]: Yes and no, It is nice to have it as a method, but then if we rely on the static methods which gives the defaults, then it will be nightmare to find all the defaults. So I would rather have the defaults like this, and in the last user exposed place (which is redundant), so user can see at one glance the whole defaults.

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 23 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: done :)

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 25 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: Hmm.... I would say, these `mComponentNames` should be consistent among ranks (I remember your use case), but woudn't it be the responsibility of the person who gives this as input to make them consistent across ranks before providing it?

Otherwise, aren't we mixing responsibilities if we integrate communication for `mComponentNames` in this?

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 26 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: done :)

File: applications/HDF5Application/custom_io/hdf5_file.cpp (hunk 28 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_file.cpp]: done :)

File: applications/HDF5Application/custom_io/hdf5_file.cpp (hunk 29 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_file.cpp]: done :)

File: applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp (hunk 30 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_vertex_container_io.cpp:94]: I am not sure, I went through the docs, but I am not sure whether they gurantee it.

File: applications/HDF5Application/custom_python/add_custom_io_to_python.cpp (hunk 32 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_python/add_custom_io_to_python.cpp:94]: Yep, but I need it to have backward compatibility with the legacy HDF5 writing mechanism. :/

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 33 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h:358]: As @matekelemen has told, these are not exactly mesh nodes, these are some random points in the mesh used for line outputs.

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 25 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: Ok, it is passed via `Kratos::Parameters`, so we don't have to make the ctor of `ContainerComponentIO` private since, user is suppose to pass the same `Kratos::Parameters` from all ranks.

File: applications/HDF5Application/custom_io/hdf5_container_component_io.cpp (hunk 27 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_io/hdf5_container_component_io.cpp]: done :)

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 3 above)
Comment (matekelemen) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: fine, just please add a `@todo` so we can get rid of it when we drop support for `VMSAdjointElement`

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 9 above)
Comment (matekelemen) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: this one slipped under your radar

**Review by matekelemen — COMMENTED** [review]
I think this one's mostly ready, you're just missing the empty rank tests and passing the size along with array pointers.

File: applications/HDF5Application/custom_utilities/container_io_utils.h (hunk 9 above)
Comment (sunethwarna) [reply applications/HDF5Application/custom_utilities/container_io_utils.h]: done :)

**Comment by sunethwarna** [comment]
Finally, I have added a zero rank test as well @matekelemen  :)

**Review by matekelemen — DISMISSED** [review]
Thanks!

**Comment by matekelemen** [comment]
I can't reproduce the failures locally.

**Review by matekelemen — APPROVED** [review]
image
